src/python/pants/rules/core/strip_source_roots.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import cast
+
+from pants.build_graph.files import Files
+from pants.engine.fs import EMPTY_SNAPSHOT, Digest, DirectoryWithPrefixToStrip, Snapshot
+from pants.engine.legacy.graph import HydratedTarget
+from pants.engine.rules import rule, subsystem_rule
+from pants.engine.selectors import Get
+from pants.source.source_root import NoSourceRootError, SourceRootConfig
+
+
+@dataclass(frozen=True)
+class SourceRootStrippedSources:
+  """Wrapper for a snapshot of targets whose source roots have been stripped."""
+  snapshot: Snapshot
+
+
+@dataclass(frozen=True)
+class StripSourceRootsRequest:
+  """A request to strip source roots for every file in the snapshot.
+
+  The field `sentinel_file` is used to determine the source root for the files to be stripped. This
+  assumes that every file shares the same source root, which should be true in practice as the
+  `sources` field for a target always has files in the same source root (outside of aggregate
+  targets, which we don't expect to ever strip the source root from). See
+  https://github.com/pantsbuild/pants/pull/9112#discussion_r377999025 for more context on this
+  design.
+  """
+  snapshot: Snapshot
+  sentinel_file: str
+
+  def determine_source_root(self, *, source_root_config: SourceRootConfig) -> str:
+    source_roots_object = source_root_config.get_source_roots()
+    source_root = source_roots_object.safe_find_by_path(self.sentinel_file)
+    if source_root is not None:
+      return cast(str, source_root.path)
```

Comment:
What is the advantage of `cast` over `str(source_root.path)`? And also, why would it ever not be a string to begin with?

Reply:
`cast` has no runtime effect, it's only used by MyPy. Here, MyPy infers that `source_root.path` has type `Any` because `source_roots.py` is not properly typed yet. 

`cast` has the effect of saying "trust me on this one: this is a `str`". The neat thing about using `cast` is that once we do add type hints to `source_roots.py` so that MyPy can correctly infer that `source_root.path` is indeed a `str`, then MyPy will warn that we no longer need this `cast` and we can remove it. It's a neat way to get incremental type hints without having to rewrite the world all at once.